### PR TITLE
Data Generator

### PR DESCRIPTION
**Pull request by vessemer**
The data augmentation process is a very common stage for a plenty of approaches in machine learning, especially in computer vision. The majority of the solutions of the #1, #2, #3 has the similar augmentation pipeline (e.g. [grt123](https://github.com/WGierke/concept-to-clinic/blob/457b7a7528fda28ed556d234135d48595eaa13ce/docs/algorithm_grt123.md), [Therapixel](https://github.com/WGierke/concept-to-clinic/blob/d9a20abfb817642ba6d98e403ed6659446125853/docs/algorithm_therapixel.md), [Daniel Hammack](https://github.com/WGierke/concept-to-clinic/blob/a47d8f28435f612d93a8eee694ac5c82500e710f/docs/algorithm_hammack.md), etc.). The aforementioned pipeline mainly consists of Hounsfield scaling, lungs segmentation (#120), CT re-orientation and a batch of trivial, yet resource consumptive spacial operations: zoom, rotation, shear, shift, flip and combinations of them. The convenient solution for the latter process is to build a generator which will yield a processed patches via affine transformations. The worth notable example of such a generator is [Keras DataGenerator](https://keras.io/preprocessing/image/) complemented by the [thorough description](https://blog.keras.io/building-powerful-image-classification-models-using-very-little-data.html). The main advantage of the Keras approach is that it for each patch build new transformation matrix from scratch but apply it only ones in comparison with step-by-step application of `scipy.ndimage` perturbations e.g. [code of Daniel Hammack](https://github.com/dhammack/DSB2017/blob/67f6474768fa47f9be95c31aa41609e0a986d8a2/training_code/aws/data_generator_fn3.py#L126-L128) where only for a rotation around the origin two calls of affine transformations were performed. Unfortunately, the main disadvantage for us is that Keras DataGenerator doesn't work with 3D data. Thus what have I done is the extension of the DataGenerator in order to deal with volumetric data such as CT scans. I've also achieved compatibility with `scipy.ndimage` to check integrity of the methods by the trusted functions (i.e. `scipy.ndimage.zoom`, `.rotate`, etc.) which I further have used in the unit tests. Resulted speedup is about 2.5 in comparison with sequantial calls of `scipy.ndimage` functions.

### Random Rotation
rand_rotation_b
rand_rotation_full_b

### Random Zoom 
rand_zoom_b

### Random Shear 
random_shear

### Random Shift
rand_shift_b

### Agregated
one
two

## CLA
- [x] I have signed the CLA; if other committers are in the commit history, they have signed the CLA as well

### CODE REVIEW

File: prediction/src/tests/test_load_dicom.py [hunk 1]
```diff
@@ -5,8 +5,8 @@
 import numpy as np
 import pytest
 
-from ..preprocess import errors
```
Comment (lamby) [inline prediction/src/tests/test_load_dicom.py]: How come?

File: prediction/src/preprocess/preprocess_ct.py [hunk 2]
```diff
@@ -82,7 +82,9 @@ def __init__(self, params=None):
 
     def __call__(self, voxel_data, meta):
         if not isinstance(meta, load_ct.MetaData):
-            raise ValueError('The meta should be an instance of %s.' % str(load_ct.MetaData))
+            print(meta.__class__)
```
Comment (lamby) [inline prediction/src/preprocess/preprocess_ct.py]: Whoops? :)

File: prediction/src/preprocess/load_ct.py [hunk 3]
```diff
@@ -127,14 +127,26 @@ def extract_spacing_dcm(self):
...
+            return [spacing[1], spacing[0], slice_thickness]
+        else:
+            return [slice_thickness, spacing[0], spacing[1]]
 
     def extract_spacing_mhd(self):
-        # Again, taking into account ijk -> xyz transformation
-        return list(reversed(self.meta.GetSpacing()))
+        if self.xyz_order:
+            return list(reversed(self.meta.GetSpacing()))
+        else:
```
Comment (lamby) [inline prediction/src/preprocess/load_ct.py]: No need for an else; you've just called ``return`` on the previous lin. (and elsewhere)

File: prediction/src/preprocess/crop_patches.py [hunk 4]
```diff
@@ -0,0 +1,80 @@
...
+    Yields:
+        np.ndarray: a cropped patch from the CT scan.
+    """
+    patch = crop_patch(ct_array, patch_shape, centroids, meta)
+    patches = list()
+    while True:
+        try:
+            patches.append(next(patch))
+        except StopIteration:
+
```
Comment (lamby) [inline prediction/src/preprocess/crop_patches.py]: flake might complain here, not sure...

Comment (vessemer) [reply prediction/src/preprocess/crop_patches.py]: Looks ugly, agreed. I've replaced it via `itertools.islice`.

**Comment by vessemer** [comment]
@lamby, all rebased and stashed.

**Comment by reubano** [comment]
I'm having trouble figuring out what issue this addresses. Ideas @vessemer @lamby?

**Comment by reubano** [comment]
Created #137 to accompany this
